gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -295,24 +304,100 @@ pub enum ClnRpcError {
 struct ClnHtlcInterceptor {
     subscriptions:
         Mutex<HashMap<u64, mpsc::Sender<Result<SubscribeInterceptHtlcsResponse, Status>>>>,
+    outcomes: Mutex<HashMap<u64, oneshot::Sender<Action>>>,
 }
 
 impl ClnHtlcInterceptor {
     fn new() -> Self {
         Self {
             subscriptions: Mutex::new(HashMap::new()),
+            outcomes: Mutex::new(HashMap::new()),
+        }
+    }
+
+    async fn intercept_htlc(&self, payload: HtlcAccepted) -> serde_json::Value {
+        let short_channel_id = payload.onion.short_channel_id;
+        let htlc_expiry = payload.htlc.cltv_expiry;
+
+        if let Some(subscription) = self.subscriptions.lock().await.get(&short_channel_id) {
+            let payment_hash = payload.htlc.payment_hash.to_vec();
+
+            match subscription
+                .send(Ok(SubscribeInterceptHtlcsResponse {
+                    payment_hash: payment_hash.clone(),
+                    incoming_amount_msat: payload.htlc.amount.msats,
+                    outgoing_amount_msat: payload.onion.forward_amount.msats,
+                    incoming_expiry: htlc_expiry,
+                    short_channel_id,
+                    // TODO: generate unique id for each intercepted HTLC
+                    intercepted_htlc_id: payment_hash.clone(),
+                }))
+                .await
+            {
+                Ok(_) => {
+                    // Open a channel to receive the outcome of the HTLC processing
+                    let (sender, receiver) = oneshot::channel::<Action>();
+                    self.outcomes.lock().await.insert(short_channel_id, sender);
+
+                    // If the gateway does not respond within the HTLC expiry,
+                    // Automatically respond with a failure message.
+                    return tokio::time::timeout(
+                        Duration::from_secs(5),
+                        self.await_htlc_processing(receiver, payment_hash.clone()),
+                    )
```

Comment:
TODO: don't block here, HTLCs can be processed in parallel imo. Not urgent though.

Reply:
Marked with a TODO.

I didn't yet look into how to make CLN process HTLCs in parallel. This is a simple migration from our `htlc_accepted` hook pattern, where we block while processing every HTLC intercepted